designs/auto-certs-renewal.md

```diff
@@ -0,0 +1,122 @@
+
+1. Add rolloutBefore.certificatesExpiryDays field to [EtcdadmCluster](https://github.com/aws/etcdadm-controller/blob/main/api/v1beta1/etcdadmcluster_types.go#L38), exposing and passing it from EKSA. The business logic of this field and reconciler should be:
+    1. apply machine annotation `machine.cluster.x-k8s.io/certificates-expiry` during etcd node creation.
+    2. check if roll out is needed. Refer https://github.com/kubernetes-sigs/cluster-api/blob/main/controlplane/kubeadm/internal/controllers/controller.go#L380.
+    3. if roll out is needed for any etcd node, upgrade all etcd nodes.
+
+However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.
+
+### Cluster Yaml Spec Example
+
+```
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: Cluster
+spec:
+  certificatesConfiguration:
+    rolloutBeforeExpiry: 
+      // whether to rollout etcd and cp nodes before certs expiring
+      enabled: true/false (false by default, auto certs renew require opt-in) 
+      // allow adding finer control in the future
+```
+
+### Tech Details
+
+1. `certificatesConfiguration.rolloutBeforeExpiry.enabled` should be true when user run “eksctl anywhere generate clusterconfig”
+2. When `rolloutBeforeExpiry` is enabled, the cp/etcd nodes are rollout when certs are expiring in **7** days
+3. etcd certs renewal has a side-effect of renewing cp certs (etcd-rollout → etcd api change → cp rollout → cp certs renewal). We should only renew etcd certs to avoid unnecessary cp rollout.
+4. ETCD cluster rollout should undergo the same process as upgrading cluster, making sure that etcd cluster is accessible from cp, aka, no downtime for the k8s cluster.
+
+## Alternative Solutions
+
+### Cronjob
+
+The script run by the job would need to be tailored/tested for each provider/OS combinations. It’s error prone and hard to develop/maintain. Manually renew certs in bottlerocket is very tricky. Instead of running the binary locally, you have to pull the etcdadm/kubeadm image locally, mount folders, configuring container runtime, calling bottlerocket api etc to make it work. Besides, when there is bug or dependencies change, you would perhap need to test the fix against multiple environments.
```

Comment:
These points are really interesting, thanks for noting them. To be able to think critically about the alternatives I think we need a little elaboration on the 'why' and a little more explanation on the what. Where does this CronJob run? Why is it error prone? Why is renewing certs in BottleRocket tricky?

Reply:
added